### ISSUE TITLE
Race condition in task-graph can cause false success

### ISSUE THREAD

**Issue by stuart-stanley** [RackHD/RackHD#295]
If you have a graph that has multiple terminus tasks that depend on the same waitOn like the following: (create-switch-snap-pollers, update-node-name, and update-lookups all waitOn 'catalog-snap')
{{ tasks: [
    { label: 'ping-host', taskName: 'Task.Snmp.Ping' },
    {
      label: 'collect-snmp',
      taskName: 'Task.Snmp.Collect.Discovery',
      waitOn:
        { 'ping-host' : 'succeeded' }
    },
    {
      label: 'catalog-snmp',
      taskName: 'Task.Snmp.Catalog',
      waitOn:
        { 'ping-host': 'succeeded' }
    },
    {
      label: 'update-lookups',
      taskName: 'Task.Snmp.Update.Lookups',
      waitOn:
        { 'catalog-snmp': 'succeeded' }
    },
    {
      label: 'create-switch-snmp-pollers',
      taskName: 'Task.Pollers.CreateDefault',
      waitOn:
        { 'catalog-snmp': 'succeeded' }
    },
    {
      label: 'update-node-name',
      taskName: 'Task.Snmp.Node.Update',
      waitOn:
        { 'catalog-snmp'': 'succeeded' }
    }
]
};}}
If one of those terminus tasks fail and it is the last of the three to finish, the graph fails.
If one of those terminus tasks fail and isn't the last of the three to finish, the graph is marked as a success.
Basically, the entire graph is marked with the status of the last one that completes...

**Comment by benbp** [comment]
Thanks @stuart-stanley! This is a thorny one :) Will take a look. I bet it might have to do something with us just checking if all other tasks are "done" when a task finishes, in order to decide whether to complete the graph, rather than if all other tasks have completed successfully, and the race condition happens when we succeed the graph before we fail it (a call to fail the graph after we have marked it succeeded will basically be a noop). FYI @VulpesArtificem

**Comment by stuart-stanley** [comment]
@benbp if you think this one is thorny, wait till you look at 290 :)

**Comment by benbp** [comment]
I was able to reproduce this issue in a more simple case with this graph:

```
    NoOpJob.prototype._run = function run() {
        var self = this;
        logger.info("RUNNING NOOP JOB");
        Promise.delay(this.delay).then(function() {
            if (self.options.fail) {
                self._done(new Error("Injected failure"));
                return;
            }
            self._done();
        });
    };
```

And changing the `_run` function for on-tasks/lib/jobs/noop-job.js to 
```
    NoOpJob.prototype._run = function run() {
        var self = this;
        logger.info("RUNNING NOOP JOB");
        Promise.delay(this.delay).then(function() {
            if (self.options.fail) {
                self._done(new Error("Injected failure"));
                return;
            }
            self._done();
        });
    };
```
